tracing/span.go

```diff
@@ -366,35 +342,138 @@ func (s *Span) createChild(name string, spanType SpanType, component string) *Sp
-		return nil, &InvalidSpanTypeError{SpanTypeServer, s.spanType}
+type contextKey int
+
+const (
+	serverSpanKey contextKey = iota
+	childSpanKey
+)
+
+// FlagMask values.
+//
+// Reference: https://github.com/reddit/baseplate.py/blob/1ca8488bcd42c8786e6a3db35b2a99517fd07a99/baseplate/observers/tracing.py#L60-L64
+const (
+	// When set, this trace passes all samplers.
+	FlagMaskDebug int64 = 1
+)
+
+const (
+	counterKeyPrefix   = "counter."
+	baseplateComponent = "baseplate"
+)
+
+type baseSpan struct {
+	trace *Trace
+	hooks *SpanHooks
+
+	name string
+}
+
+func (s baseSpan) Name() string {
+	return s.name
+}
+
+func (s baseSpan) Trace() *Trace {
+	return s.trace
+}
+
+func (s baseSpan) Hooks() *SpanHooks {
+	return s.hooks
+}
+
+func (s baseSpan) SpanType() SpanType {
+	return SpanTypeUnknown
+}
+
+func (s baseSpan) Start() {
+	return
+}
+
+func (s baseSpan) Stop(err error) {
+	return
+}
+
+func (s baseSpan) LogError(msg string, err error) {
+	if s.Trace().tracer.Logger != nil {
+		s.Trace().tracer.Logger("OnStart hook error: " + err.Error())
 	}
-	return context.WithValue(ctx, serverSpanKey, s), nil
 }
 
-func getSpanFromContext(ctx context.Context, key contextKey) *Span {
-	if s := ctx.Value(key); s != nil {
-		if span, ok := s.(*Span); ok {
-			return span
-		}
+// ServerSpan defines a tracing span.
+type ServerSpan struct {
+	baseSpan
+}
+
+// SpanType returns the SpanType for s.
+func (s ServerSpan) SpanType() SpanType {
+	return SpanTypeServer
+}
+
+// Stop ends the span
+func (s ServerSpan) Stop(err error) {
+	if err != nil && errors.Is(err, context.DeadlineExceeded) {
+		s.Trace().setTag(ZipkinBinaryAnnotationKeyTimeOut, true)
 	}
-	return nil
+}
+
+// SetServerSpan sets the span as the ServerSpan on the given context
+func (s ServerSpan) SetServerSpan(ctx context.Context) context.Context {
+	return context.WithValue(ctx, serverSpanKey, &s)
+}
+
+// ClientSpan defines a tracing span for a 'client' call.
+type ClientSpan struct {
+	baseSpan
+}
+
+// SpanType returns the SpanType for s.
+func (s ClientSpan) SpanType() SpanType {
+	return SpanTypeClient
+}
+
+// LocalSpan defines a tracing span for a 'local' call.
+type LocalSpan struct {
+	baseSpan
+
+	component string
+}
+
+// Stop ends the span
+func (s LocalSpan) Stop(err error) {
+	if s.component != "" {
+		s.Trace().setTag(ZipkinBinaryAnnotationKeyLocalComponent, s.component)
+	}
+}
+
+// SpanType returns the SpanType for s.
+func (s LocalSpan) SpanType() SpanType {
+	return SpanTypeLocal
 }
 
 // GetServerSpan gets the ServerSpan from the given context
-func GetServerSpan(ctx context.Context) *Span {
-	return getSpanFromContext(ctx, serverSpanKey)
+func GetServerSpan(ctx context.Context) *ServerSpan {
+	if span, ok := ctx.Value(serverSpanKey).(*ServerSpan); ok {
+		return span
+	}
+	return nil
 }
 
-// GetChildSpan gets the ChildSpan from the given context
+// GetChildSpan gets the child span from the given context
 func GetChildSpan(ctx context.Context) *Span {
```

Comment:
Since `Span` is an interface now, it shouldn't be a pointer.

Reply:
I wasn't sure about how to handle that since I can't return `nil` if it's _not_ a pointer?  Do I just return an empty one?  Or maybe make a "null" span that doesn't do anything?

Reply:
interfaces are pointers. you can return nil for an interface.